File: test/lib/vbrain/routines_test.rb
require "test_helper"
require "vbrain"

class RoutinesTest < Minitest::Test
  FIXED_NOW = Time.utc(2026, 5, 28, 12, 0, 0).freeze

  def setup
    @original_tz = ENV["TZ"]
    ENV["TZ"] = "UTC"
  end

  def teardown
    ENV["TZ"] = @original_tz
  end

  def test_load_all_returns_empty_when_config_missing
    with_isolated_data_home do |_|
      assert_equal [], VBrain::Routines.load_all
      assert_equal [], VBrain::Routines.enabled
      assert_nil VBrain::Routines.find("anything")
    end
  end

  def test_add_creates_yaml_and_returns_entry_with_next_run
    with_isolated_data_home do |_|
      entry = VBrain::Routines.add!(
        slug:        "morning-brief",
        description: "Resumo da manhã",
        schedule:    "0 6 * * *",
        prompt:      "Liste emails INBOX e reuniões de hoje.",
        now:         FIXED_NOW
      )

      assert_equal "morning-brief", entry["slug"]
      assert_equal "0 6 * * *",     entry["schedule"]
      assert_equal "2026-05-29T06:00:00Z", entry["next_run"]
      assert_nil   entry["last_run"]
      assert_equal true, entry["enabled"]

      reloaded = VBrain::Routines.load_all
      assert_equal 1, reloaded.size
      assert_equal "2026-05-29T06:00:00Z", reloaded.first["next_run"]
    end
  end

  def test_add_allows_nil_schedule_for_manual_only_routines
    with_isolated_data_home do |_|
      entry = VBrain::Routines.add!(slug: "manual", description: "", prompt: "x")
      assert_nil entry["schedule"]
    end
  end

  def test_add_rejects_invalid_cron_expression
    with_isolated_data_home do |_|
      assert_raises(VBrain::Routines::Error) do
        VBrain::Routines.add!(slug: "x", description: "", prompt: "p", schedule: "every hour")
      end
      assert_raises(VBrain::Routines::Error) do
        VBrain::Routines.add!(slug: "y", description: "", prompt: "p", schedule: "0 6 * *")
      end
    end
  end

  def test_add_normalizes_slug_via_Slug_from
    with_isolated_data_home do |_|
      entry = VBrain::Routines.add!(
        slug:        "Morning Brief!",
        description: "x",
        prompt:      "y"
      )
      assert_equal "morning-brief", entry["slug"]
    end
  end

  def test_add_rejects_empty_slug_and_prompt
    with_isolated_data_home do |_|
      assert_raises(VBrain::Routines::Error) do
        VBrain::Routines.add!(slug: "", description: "", prompt: "x")
      end
      assert_raises(VBrain::Routines::Error) do
        VBrain::Routines.add!(slug: "valid", description: "", prompt: "")
      end
    end
  end

  def test_add_collision_raises_without_replace
    with_isolated_data_home do |_|
      VBrain::Routines.add!(slug: "x", description: "", prompt: "first")
      err = assert_raises(VBrain::Routines::Error) do
        VBrain::Routines.add!(slug: "x", description: "", prompt: "second")
      end
      assert_includes err.message, "already exists"
    end
  end

  def test_add_replace_updates_in_place_preserving_order
    with_isolated_data_home do |_|
      VBrain::Routines.add!(slug: "a", description: "", prompt: "p1")
      VBrain::Routines.add!(slug: "b", description: "", prompt: "p2")
      VBrain::Routines.add!(slug: "c", description: "", prompt: "p3")
      VBrain::Routines.add!(slug: "b", description: "novo", prompt: "p2-new", replace: true)

      slugs = VBrain::Routines.load_all.map { |r| r["slug"] }
      assert_equal %w[a b c], slugs
      b = VBrain::Routines.find("b")
      assert_equal "novo",   b["description"]
      assert_equal "p2-new", b["prompt"]
    end
  end

  def test_enabled_filters_disabled_routines
    with_isolated_data_home do |_|
      VBrain::Routines.add!(slug: "on1", description: "", prompt: "x")
      VBrain::Routines.add!(slug: "off", description: "", prompt: "x", enabled: false)
      VBrain::Routines.add!(slug: "on2", description: "", prompt: "x")

      slugs = VBrain::Routines.enabled.map { |r| r["slug"] }
      assert_equal %w[on1 on2], slugs
    end
  end

  def test_remove_returns_true_when_present_and_false_otherwise
    with_isolated_data_home do |_|
      VBrain::Routines.add!(slug: "x", description: "", prompt: "p")
      assert_equal true,  VBrain::Routines.remove!("x")
      assert_equal false, VBrain::Routines.remove!("x")
      assert_nil VBrain::Routines.find("x")
    end
  end

  def test_load_all_treats_missing_enabled_as_true
    with_isolated_data_home do |dir|
      FileUtils.mkdir_p(File.dirname(VBrain::Routines.config_path))
      File.write(VBrain::Routines.config_path, YAML.dump(
        "routines" => [{ "slug" => "x", "description" => "", "prompt" => "p" }]
      ))
      assert_equal true, VBrain::Routines.find("x")["enabled"]
    end
  end

  def test_compute_next_run_is_deterministic
    n1 = VBrain::Routines.compute_next_run("0 6 * * *", FIXED_NOW)
    n2 = VBrain::Routines.compute_next_run("0 6 * * *", FIXED_NOW)
    assert_equal n1.iso8601, n2.iso8601
    assert_equal "2026-05-29T06:00:00Z", n1.utc.iso8601
  end

  def test_compute_next_run_handles_hourly_and_weekly
    # 2026-05-28 = Thursday (wday=4). "0 10 * * 3" = every Wednesday 10:00.
    hourly  = VBrain::Routines.compute_next_run("0 * * * *", FIXED_NOW)
    weekly  = VBrain::Routines.compute_next_run("0 10 * * 3", FIXED_NOW)
    assert_equal "2026-05-28T13:00:00Z", hourly.utc.iso8601
    assert_equal "2026-06-03T10:00:00Z", weekly.utc.iso8601
  end

  def test_claim_due_returns_nothing_when_no_routines
    with_isolated_data_home do |_|
      assert_equal [], VBrain::Routines.claim_due!(now: FIXED_NOW)
    end
  end

  def test_claim_due_returns_routines_past_next_run_and_advances_them
    with_isolated_data_home do |_|
      VBrain::Routines.add!(slug: "h", description: "", prompt: "p",
                            schedule: "0 * * * *", now: FIXED_NOW)
      VBrain::Routines.add!(slug: "d", description: "", prompt: "p",
                            schedule: "0 6 * * *", now: FIXED_NOW)

      # 1 hour later: hourly is due, daily-06 is not.
      one_hour_later = FIXED_NOW + 3600
      due = VBrain::Routines.claim_due!(now: one_hour_later)
      assert_equal ["h"], due.map { |r| r["slug"] }

      reloaded_h = VBrain::Routines.find("h")
      assert_equal one_hour_later.iso8601, reloaded_h["last_run"]
      # next_run advances to the next hourly tick after one_hour_later
      assert_equal "2026-05-28T14:00:00Z", reloaded_h["next_run"]

      reloaded_d = VBrain::Routines.find("d")
      assert_nil reloaded_d["last_run"]
      assert_equal "2026-05-29T06:00:00Z", reloaded_d["next_run"]
    end
  end

  def test_claim_due_returns_previous_last_run_in_due_entry
    with_isolated_data_home do |_|
      VBrain::Routines.add!(slug: "h", description: "", prompt: "p",
                            schedule: "0 * * * *", now: FIXED_NOW)

      # First claim: no prior last_run.
      first_tick = FIXED_NOW + 3600
      first = VBrain::Routines.claim_due!(now: first_tick)
      assert_equal 1, first.size
      assert_nil first.first["last_run"],
                 "due entry should expose the previous last_run (nil before any run)"

      # Second claim: prior last_run is the first tick.
      second_tick = first_tick + 3600
      second = VBrain::Routines.claim_due!(now: second_tick)
      assert_equal 1, second.size
      assert_equal first_tick.iso8601, second.first["last_run"],
                   "due entry should expose the previous last_run, not the just-claimed one"
    end
  end

  def test_claim_due_is_idempotent_when_called_twice_in_same_tick
    with_isolated_data_home do |_|
      VBrain::Routines.add!(slug: "h", description: "", prompt: "p",
                            schedule: "0 * * * *", now: FIXED_NOW)
      later = FIXED_NOW + 3600
      first  = VBrain::Routines.claim_due!(now: later)
      second = VBrain::Routines.claim_due!(now: later)
      assert_equal 1, first.size
      assert_equal 0, second.size
    end
  end

  def test_claim_due_skips_disabled
    with_isolated_data_home do |_|
      VBrain::Routines.add!(slug: "off", description: "", prompt: "p",
                            schedule: "0 * * * *", enabled: false, now: FIXED_NOW)
      due = VBrain::Routines.claim_due!(now: FIXED_NOW + 7200)
      assert_equal [], due.map { |r| r["slug"] }
    end
  end

  def test_claim_due_skips_routines_without_schedule
    with_isolated_data_home do |_|
      VBrain::Routines.add!(slug: "manual", description: "", prompt: "p")
      assert_equal [], VBrain::Routines.claim_due!(now: FIXED_NOW + 86_400)
    end
  end

  def test_claim_due_backfills_next_run_when_missing
    with_isolated_data_home do |_|
      FileUtils.mkdir_p(File.dirname(VBrain::Routines.config_path))
      File.write(VBrain::Routines.config_path, YAML.dump(
        "routines" => [{
          "slug" => "x", "description" => "", "prompt" => "p",
          "schedule" => "0 * * * *", "enabled" => true
        }]
      ))
      # First claim with no next_run set: routine is "due" if computed
      # next_run falls before our now. Since fugit.next_time(now) is always
      # strictly after now, the routine should NOT be due, but next_run
      # should be backfilled.
      due = VBrain::Routines.claim_due!(now: FIXED_NOW)
      assert_equal [], due.map { |r| r["slug"] }
      r = VBrain::Routines.find("x")
      assert_equal "2026-05-28T13:00:00Z", r["next_run"]
    end
  end
end

File: internal/routines/routines_test.go
package routines_test

import (
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/virtual360-io/vbrain/internal/routines"
)

var fixedNow = time.Date(2026, 5, 28, 12, 0, 0, 0, time.UTC)

func isolate(t *testing.T) {
	t.Helper()
	t.Setenv("VBRAIN_HOME", t.TempDir())
}

func sp(s string) *string { return &s }

// add é um wrapper enxuto pros testes.
func add(t *testing.T, slug, desc, prompt string, schedule *string, enabled bool) (routines.Routine, error) {
	t.Helper()
	return routines.Add(slug, desc, prompt, schedule, enabled, false, fixedNow)
}

func TestLoadAllEmptyWhenMissing(t *testing.T) {
	isolate(t)
	all, _ := routines.LoadAll()
	if len(all) != 0 {
		t.Fatalf("LoadAll = %v", all)
	}
	en, _ := routines.Enabled()
	if len(en) != 0 {
		t.Fatalf("Enabled = %v", en)
	}
	if r, _ := routines.Find("anything"); r != nil {
		t.Fatalf("Find = %v", r)
	}
}

func TestAddCreatesYamlWithNextRun(t *testing.T) {
	isolate(t)
	e, err := add(t, "morning-brief", "Resumo da manhã", "Liste emails.", sp("0 6 * * *"), true)
	if err != nil {
		t.Fatal(err)
	}
	if e.Slug != "morning-brief" || e.Schedule == nil || *e.Schedule != "0 6 * * *" {
		t.Fatalf("entry = %+v", e)
	}
	if e.NextRun == nil || *e.NextRun != "2026-05-29T06:00:00Z" {
		t.Fatalf("next_run = %v", e.NextRun)
	}
	if e.LastRun != nil || !e.Enabled {
		t.Fatalf("last_run/enabled = %v/%v", e.LastRun, e.Enabled)
	}
	all, _ := routines.LoadAll()
	if len(all) != 1 || all[0].NextRun == nil || *all[0].NextRun != "2026-05-29T06:00:00Z" {
		t.Fatalf("reload = %+v", all)
	}
}

func TestAddAllowsNilSchedule(t *testing.T) {
	isolate(t)
	e, err := routines.Add("manual", "", "x", nil, true, false, fixedNow)
	if err != nil {
		t.Fatal(err)
	}
	if e.Schedule != nil {
		t.Fatalf("schedule = %v, want nil", e.Schedule)
	}
}

func TestAddRejectsInvalidCron(t *testing.T) {
	isolate(t)
	if _, err := add(t, "x", "", "p", sp("every hour"), true); err == nil {
		t.Error("'every hour' deveria falhar")
	}
	if _, err := add(t, "y", "", "p", sp("0 6 * *"), true); err == nil {
		t.Error("'0 6 * *' (4 campos) deveria falhar")
	}
}

func TestAddNormalizesSlug(t *testing.T) {
	isolate(t)
	e, err := add(t, "Morning Brief!", "x", "y", nil, true)
	if err != nil {
		t.Fatal(err)
	}
	if e.Slug != "morning-brief" {
		t.Fatalf("slug = %q", e.Slug)
	}
}

func TestAddRejectsEmptySlugAndPrompt(t *testing.T) {
	isolate(t)
	if _, err := add(t, "", "", "x", nil, true); err == nil {
		t.Error("slug vazio deveria falhar")
	}
	if _, err := add(t, "valid", "", "", nil, true); err == nil {
		t.Error("prompt vazio deveria falhar")
	}
}

func TestAddCollisionRaisesWithoutReplace(t *testing.T) {
	isolate(t)
	add(t, "x", "", "first", nil, true)
	if _, err := add(t, "x", "", "second", nil, true); err == nil {
		t.Fatal("colisão deveria falhar sem replace")
	}
}

func TestAddReplaceUpdatesInPlacePreservingOrder(t *testing.T) {
	isolate(t)
	add(t, "a", "", "p1", nil, true)
	add(t, "b", "", "p2", nil, true)
	add(t, "c", "", "p3", nil, true)
	if _, err := routines.Add("b", "novo", "p2-new", nil, true, true, fixedNow); err != nil {
		t.Fatal(err)
	}
	all, _ := routines.LoadAll()
	if len(all) != 3 || all[0].Slug != "a" || all[1].Slug != "b" || all[2].Slug != "c" {
		t.Fatalf("ordem = %+v", all)
	}
	b, _ := routines.Find("b")
	if b.Description != "novo" || b.Prompt != "p2-new" {
		t.Fatalf("b = %+v", b)
	}
}

func TestEnabledFiltersDisabled(t *testing.T) {
	isolate(t)
	add(t, "on1", "", "x", nil, true)
	add(t, "off", "", "x", nil, false)
	add(t, "on2", "", "x", nil, true)
	en, _ := routines.Enabled()
	if len(en) != 2 || en[0].Slug != "on1" || en[1].Slug != "on2" {
		t.Fatalf("enabled = %+v", en)
	}
}

func TestRemove(t *testing.T) {
	isolate(t)
	add(t, "x", "", "p", nil, true)
	if ok, _ := routines.Remove("x"); !ok {
		t.Error("remove deveria retornar true")
	}
	if ok, _ := routines.Remove("x"); ok {
		t.Error("segundo remove deveria retornar false")
	}
}

func TestLoadAllTreatsMissingEnabledAsTrue(t *testing.T) {
	isolate(t)
	os.MkdirAll(filepath.Dir(routines.ConfigPath()), 0o755)
	os.WriteFile(routines.ConfigPath(), []byte("routines:\n  - slug: x\n    description: \"\"\n    prompt: p\n"), 0o644)
	r, _ := routines.Find("x")
	if r == nil || !r.Enabled {
		t.Fatalf("enabled ausente deveria virar true: %+v", r)
	}
}

func TestComputeNextRunIsDeterministic(t *testing.T) {
	n1, _ := routines.ComputeNextRun("0 6 * * *", fixedNow)
	n2, _ := routines.ComputeNextRun("0 6 * * *", fixedNow)
	if !n1.Equal(n2) {
		t.Fatal("não determinístico")
	}
	if got := n1.Format(time.RFC3339); got != "2026-05-29T06:00:00Z" {
		t.Fatalf("got %q", got)
	}
}

func TestComputeNextRunHourlyAndWeekly(t *testing.T) {
	hourly, _ := routines.ComputeNextRun("0 * * * *", fixedNow)
	weekly, _ := routines.ComputeNextRun("0 10 * * 3", fixedNow)
	if got := hourly.Format(time.RFC3339); got != "2026-05-28T13:00:00Z" {
		t.Errorf("hourly = %q", got)
	}
	if got := weekly.Format(time.RFC3339); got != "2026-06-03T10:00:00Z" {
		t.Errorf("weekly = %q", got)
	}
}

func TestClaimDueEmpty(t *testing.T) {
	isolate(t)
	due, _ := routines.ClaimDue(fixedNow)
	if len(due) != 0 {
		t.Fatalf("due = %v", due)
	}
}

func TestClaimDuePastNextRunAdvances(t *testing.T) {
	isolate(t)
	add(t, "h", "", "p", sp("0 * * * *"), true)
	add(t, "d", "", "p", sp("0 6 * * *"), true)

	oneHourLater := fixedNow.Add(time.Hour)
	due, _ := routines.ClaimDue(oneHourLater)
	if len(due) != 1 || due[0].Slug != "h" {
		t.Fatalf("due = %+v", due)
	}
	h, _ := routines.Find("h")
	if h.LastRun == nil || *h.LastRun != "2026-05-28T13:00:00Z" {
		t.Errorf("h.last_run = %v", h.LastRun)
	}
	if h.NextRun == nil || *h.NextRun != "2026-05-28T14:00:00Z" {
		t.Errorf("h.next_run = %v", h.NextRun)
	}
	d, _ := routines.Find("d")
	if d.LastRun != nil || d.NextRun == nil || *d.NextRun != "2026-05-29T06:00:00Z" {
		t.Errorf("d = %+v", d)
	}
}

func TestClaimDueExposesPreviousLastRun(t *testing.T) {
	isolate(t)
	add(t, "h", "", "p", sp("0 * * * *"), true)

	first := fixedNow.Add(time.Hour)
	d1, _ := routines.ClaimDue(first)
	if len(d1) != 1 || d1[0].LastRun != nil {
		t.Fatalf("primeiro claim deveria expor last_run nil: %+v", d1)
	}
	second := first.Add(time.Hour)
	d2, _ := routines.ClaimDue(second)
	if len(d2) != 1 || d2[0].LastRun == nil || *d2[0].LastRun != first.Format(time.RFC3339) {
		t.Fatalf("segundo claim deveria expor last_run anterior: %+v", d2)
	}
}

func TestClaimDueIdempotentSameTick(t *testing.T) {
	isolate(t)
	add(t, "h", "", "p", sp("0 * * * *"), true)
	later := fixedNow.Add(time.Hour)
	d1, _ := routines.ClaimDue(later)
	d2, _ := routines.ClaimDue(later)
	if len(d1) != 1 || len(d2) != 0 {
		t.Fatalf("d1=%d d2=%d", len(d1), len(d2))
	}
}

func TestClaimDueSkipsDisabled(t *testing.T) {
	isolate(t)
	add(t, "off", "", "p", sp("0 * * * *"), false)
	due, _ := routines.ClaimDue(fixedNow.Add(2 * time.Hour))
	if len(due) != 0 {
		t.Fatalf("due = %+v", due)
	}
}

func TestClaimDueSkipsNoSchedule(t *testing.T) {
	isolate(t)
	add(t, "manual", "", "p", nil, true)
	due, _ := routines.ClaimDue(fixedNow.Add(24 * time.Hour))
	if len(due) != 0 {
		t.Fatalf("due = %+v", due)
	}
}

func TestClaimDueBackfillsNextRunWhenMissing(t *testing.T) {
	isolate(t)
	os.MkdirAll(filepath.Dir(routines.ConfigPath()), 0o755)
	os.WriteFile(routines.ConfigPath(), []byte("routines:\n  - slug: x\n    description: \"\"\n    prompt: p\n    schedule: 0 * * * *\n    enabled: true\n"), 0o644)

	due, _ := routines.ClaimDue(fixedNow)
	if len(due) != 0 {
		t.Fatalf("não deveria estar due (next_time é estritamente após now): %+v", due)
	}
	r, _ := routines.Find("x")
	if r.NextRun == nil || *r.NextRun != "2026-05-28T13:00:00Z" {
		t.Fatalf("next_run backfill = %v", r.NextRun)
	}
}
